### PR TITLE
Implement ReduceLROnPlateau with Backtrack

### PR DESCRIPTION
**Pull request by Jiaming-Liu**
Responding to https://github.com/pytorch/pytorch/issues/2478:
1. Enhance ReduceLROnPlateau with backtracking & autosaving function.
2. ReduceLROnPlateau.step(...) now return a status code.

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -303,3 +314,74 @@ def _init_is_better(self, mode, threshold, threshold_mode):
         else:  # mode == 'max' and epsilon_mode == 'abs':
             self.is_better = lambda a, best: a > best + threshold
             self.mode_worse = -float('Inf')
+
+
+class ReduceLROnPlateauWithBacktrack(ReduceLROnPlateau):
```
Comment (apaszke) [inline torch/optim/lr_scheduler.py]: I think it would be better to add a `backtrack` argument to `ReduceLROnPlateau`

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -249,26 +251,35 @@ def _reset(self):
...
             self.num_bad_epochs = 0
+            return self.STATUS_UPDATED_BEST
+        else:
+            if self.num_bad_epochs > self.patience:
+                self._reduce_lr(epoch)
+                self.cooldown_counter = self.cooldown
+                self.num_bad_epochs = 0
+                return self.STATUS_REDUCED_LR
+            else:
+                return self.STATUS_WAITING
```
Comment (apaszke) [inline torch/optim/lr_scheduler.py]: Not a big fan of integer return codes. I'd replace these constants with strings, so they can be used both as an enum, and can be easily printed too

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 from bisect import bisect_right
 from .optimizer import Optimizer
+from .. import load, save
```
Comment (apaszke) [inline torch/optim/lr_scheduler.py]: nit: I'd change that to `import torch` and use `torch.load` and `torch.save` to make it clear that these are not `load`, `save` methods of any class (that's what I thought when I saw them in the code)

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (Jiaming-Liu) [reply torch/optim/lr_scheduler.py]: Not sure about this -- we would need two more parameters for backtracking.

Comment (apaszke) [reply torch/optim/lr_scheduler.py]: If you need a file name make it `backtrack_filename` which enables backtracking if set, and is None otherwise (which means no backtracking)

**Comment by weiyangfb** [comment]
a rebase is needed now

**Comment by Ir1d** [comment]
Hi, any update on this? it would be great if ReduceLROnPlateau could be reloaded from a checkpoint :smile:

**Comment by yf225** [comment]
@vincentqb Would you like to check if we should move forward on this one?

**Review by vincentqb — CHANGES_REQUESTED** [review]
I would favor instead leaving the user in control of the backtracking mechanism. In particular, I would avoid managing saving/loading states in the scheduler.

I encourage the discussion around the API to continue in the original issue #2478.

**Comment by vincentqb** [comment]
Based on the comment given above, and the age of the PR, I will close it. If you would like to continue, please feel free to re-open.
